src/coins.h

```diff
@@ -130,6 +157,56 @@ struct CCoinsCacheEntry
     CCoinsCacheEntry() : flags(0) {}
     explicit CCoinsCacheEntry(Coin&& coin_) : coin(std::move(coin_)), flags(0) {}
     CCoinsCacheEntry(Coin&& coin_, unsigned char flag) : coin(std::move(coin_)), flags(flag) {}
+    ~CCoinsCacheEntry()
+    {
+        // We must always clear the flags when destroying this to remove it from
+        // the flagged linked list
+        ClearFlags();
+    }
+
+    //! Adding a flag also requires a self reference to the pair that contains
+    //! this entry in the CCoinsCache map and a reference to the head of the
+    //! flagged pair linked list.
+    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& head)
+    {
+        if (flags == 0) return;
+        if (m_flags != 0) {
+            m_flags |= flags;
+            return;
+        }
+        m_prev = &head;
+        m_next = head.second.m_next;
+        head.second.m_next = &self;
+        if (m_next != nullptr) {
+            m_next->second.m_prev = &self;
+        }
+        m_flags = flags;
+    }
+
+    inline void ClearFlags()
+    {
+        if (m_flags == 0) return;
+        m_prev->second.m_next = m_next;
+        if (m_next != nullptr) {
+            m_next->second.m_prev = m_prev;
+        }
+        m_prev = nullptr;
+        m_next = nullptr;
+        m_flags = 0;
+    }
+
+    inline uint8_t GetFlags() const { return m_flags; }
+
+    //! Get the next entry in the flagged linked list, optionally removing the
+    //! current entry and clearing the flags.
+    inline CoinsCachePair* Next(bool clear_flags = false)
```

Comment:
It's mostly convenience. Otherwise to clear flags while iterating we would have to bind `m_next` to another variable first, then call `ClearFlags` on the current pair, and then bind the the other variable holding the old `m_next` to the current pair variable.

Reply:
Can you find a way to separate the side-effects? I find this very confusing.

Reply:
Before this patch, the iteration is `it = erase ? mapCoins.erase(it) : std::next(it);`. So this is basically the same behavior, is it not?

Reply:
The std lib uses this pattern though, with `std::unordered_map::erase`. It mutates and returns next. Surely we should be following std lib conventions, no?

Reply:
I updated this to no longer clear flags.